repository/src/main/java/tech/ydb/yoj/repository/db/cache/FirstLevelCache.java

```diff
@@ -1,84 +1,211 @@
-     * transaction L1 cache.
+     * Returns a immutable copy of the entities of the specified table that are in the transaction L1 cache.
+     *
+     * @param tableDescriptor table descriptor
+     * @return unmodifiable <strong>snapshot</strong> (immutable copy) of the L1 cache for the specified table descriptor
      */
-    <E extends Entity<E>> List<E> snapshot(@NonNull Class<E> entityType);
+    <E extends Entity<E>> List<E> snapshot(@NonNull TableDescriptor<E> tableDescriptor);
 
     /**
-     * Returns a unmodifiable map containing the entities of {@code entityType} type that are in the
+     * Returns a unmodifiable map containing the entities with the specified {@code TableDescriptor} that are in the
      * transaction L1 cache. The returned map is a unmodifiable live view of transaction L1 cache;
      * changes to second affect the first. To avoid an {@link ConcurrentModificationException} exception
      * when iterating and simultaneous modification of the transaction L1 cache, you may need to make a copy.
+     *
+     * @param tableDescriptor table descriptor
+     * @return unmodifiable <strong>live view</strong> of the L1 cache for the specified table descriptor
      */
-    <E extends Entity<E>> Map<Entity.Id<E>, E> entities(@NonNull Class<E> entityType);
+    <E extends Entity<E>> Map<Entity.Id<E>, E> entities(@NonNull TableDescriptor<E> tableDescriptor);
 
-    <E extends Entity<E>> void put(@NonNull E e);
+    /**
+     * Records the entity as existing.
+     *
+     * @param tableDescriptor table descriptor
+     * @param e               existing entity
+     * @param <E>             entity type
+     * @see #putEmpty(TableDescriptor, Entity.Id)
+     */
+    <E extends Entity<E>> void put(@NonNull TableDescriptor<E> tableDescriptor, @NonNull E e);
 
-    <E extends Entity<E>> void putEmpty(@NonNull Entity.Id<E> id);
+    /**
+     * Records the entity with the specified ID as being nonexistent.
+     *
+     * @param tableDescriptor table descriptor
+     * @param id              entity ID
+     * @param <E>             entity type
+     * @see #put(TableDescriptor, Entity)
+     */
+    <E extends Entity<E>> void putEmpty(@NonNull TableDescriptor<E> tableDescriptor, @NonNull Entity.Id<E> id);
 
-    <E extends Entity<E>> boolean containsKey(@NonNull Entity.Id<E> id);
+    /**
+     * Checks whether there is an entry in the L1 cache for an entity with the specified ID (either for an entity existing, or for an entity
+     * being nonexistent).
+     *
+     * @param tableDescriptor table descriptor
+     * @param id              entity ID
+     * @param <E>             entity type
+     * @return {@code true} if the specified entity exists in the cache for the specified table descriptor; {@code false} otherwise
+     */
+    <E extends Entity<E>> boolean containsKey(@NonNull TableDescriptor<E> tableDescriptor, @NonNull Entity.Id<E> id);
+
+    // BEGIN DEPRECATED
```

Comment:
Previous `FirstLevelCache` methods that only operate on `Entity.Id`/`Entity`/`Class<Entity>` now must take a `TableDescriptor`, but these methods are used in application code (`e.g.` DNS Control Plane, but there are likely other usages). So I've added a temporary compatibility layer, to be removed after everyone migrates to the new and improved `FirstLevelCache` API that takes `TableDescriptor`s :-)

Reply:
:heavy_check_mark: Fixed via an additional layer of indirection (`FirstLevelCacheProvider`)